### ISSUE TITLE
Extras Works Differently Now?

### ISSUE THREAD

**Issue by bgile** [ktbyers/nornir_netmiko#7]
Trying to adapt all of my code to the new version 3 parameters of nornir.  Before I was able to take a connection and feed it some extra options in case I needed to change the connection type:
`task.host.open_connection("netmiko", configuration=task.nornir.config, extras={'device_type': 'cisco_ios_telnet'})`

In version 3 this no longer seems to work, I had to remove the extras portion to get it to even open a connection.  I was trying to look through both your addition of the nornir_netmiko.connections.netmiko and the nornir.core.inventory stuff but I am not quite sure how this is supposed to be handled anymore as it looks like nornir.core.inventory has not changed the open_connection method?  I also could not find any decent examples.  Hoping this is something that is probably right in front of my eyes.  Or looking for ideas on how to handle it more gracefully.  Thanks.

**Comment by ktbyers** [comment]
@bgile Typically you would pass it from inventory using `connection_options`

https://nornir.discourse.group/t/passing-enable-secret/16



Regards, 

Kirk

**Comment by bgile** [comment]
Apologize for not getting back to you faster.  Totally makes sense on where to handle it.  I recently discovered that you added an adtran_os device_type which is great!  Is there the equivalent of cisco_ios_telnet for adtran_os?  I'm trying to implement something to get telnet to work, but all I can see is there is a telnet_login method for the base class, but not a way to specify telnet.

`for host in host_list:
        host_data[host] = {
            "groups": ["adtran"],
            "connection_options": {
                "netmiko": {
                    "username": user,
                    "hostname": host",
                    "port": 23,
                    "platform": "adtran_os",
                    "extras": {
                        "device_type": "some-device-for-telnet"
                    }
                }
            }
        }`

Also, I noticed that connection_protocol is not recognized?  I was thinking about trying to feed that to extras, but I got back:
`Exception:..... TypeError: __init__() got an unexpected keyword argument 'connection_protocol'`

Thanks,

**Comment by bgile** [comment]
Looking at this a little more this might be linked somehow to my pip environment.  Local testing seems to be ok.

2020-11-16 11:47:58,126 - nornir.core.task -    ERROR -      start() - Host : task 'netmiko_send_command' failed with traceback:
Traceback (most recent call last):
  File "/home/OFFICE/bgile/.local/share/virtualenvs/adtran-vLsmWbHh/lib/python3.6/site-packages/nornir/core/task.py", line 98, in start
    r = self.task(self, **self.params)
  File "/home/OFFICE/bgile/.local/share/virtualenvs/adtran-vLsmWbHh/lib/python3.6/site-packages/nornir_netmiko/tasks/netmiko_send_command.py", line 26, in netmiko_send_command
    net_connect = task.host.get_connection(CONNECTION_NAME, task.nornir.config)
  File "/home/OFFICE/bgile/.local/share/virtualenvs/adtran-vLsmWbHh/lib/python3.6/site-packages/nornir/core/inventory.py", line 448, in get_connection
    extras=conn.extras,
  File "/home/OFFICE/bgile/.local/share/virtualenvs/adtran-vLsmWbHh/lib/python3.6/site-packages/nornir/core/inventory.py", line 499, in open_connection
    configuration=configuration,
  File "/home/OFFICE/bgile/.local/share/virtualenvs/adtran-vLsmWbHh/lib/python3.6/site-packages/nornir_netmiko/connections/netmiko.py", line 59, in open
    connection = ConnectHandler(**parameters)
  File "/home/OFFICE/bgile/.local/share/virtualenvs/adtran-vLsmWbHh/lib/python3.6/site-packages/netmiko/ssh_dispatcher.py", line 312, in ConnectHandler
    return ConnectionClass(*args, **kwargs)
  File "/home/OFFICE/bgile/.local/share/virtualenvs/adtran-vLsmWbHh/lib/python3.6/site-packages/netmiko/cisco/cisco_ios.py", line 17, in __init__
    return super().__init__(*args, **kwargs)
  File "/home/OFFICE/bgile/.local/share/virtualenvs/adtran-vLsmWbHh/lib/python3.6/site-packages/netmiko/base_connection.py", line 346, in __init__
    self._open()
  File "/home/OFFICE/bgile/.local/share/virtualenvs/adtran-vLsmWbHh/lib/python3.6/site-packages/netmiko/base_connection.py", line 351, in _open
    self.establish_connection()
  File "/home/OFFICE/bgile/.local/share/virtualenvs/adtran-vLsmWbHh/lib/python3.6/site-packages/netmiko/base_connection.py", line 910, in establish_connection
    self.telnet_login()
  File "/home/OFFICE/bgile/.local/share/virtualenvs/adtran-vLsmWbHh/lib/python3.6/site-packages/netmiko/cisco_base_connection.py", line 171, in telnet_login
    raise NetmikoAuthenticationException(msg)
netmiko.ssh_exception.NetmikoAuthenticationException: Login failed: Host

It looks like it's doing the right thing and trying to open a telnet connection but still failing for some reason.  When I execute the same code locally outside the environment it has no issues.  I am not certain if this is related to anything that is part of nornir at this time though I am confused about how pipenv would create a different outcome.  I made sure that the environment was up to date with all the libraries in its Pipfile.

**Comment by ktbyers** [comment]
Yes, there is a `adtran_os_telnet`, you would specify this as the `platform` or the `device_type` (if you specify it in `connection_options` > `extras`.

You don't need to specify the port, the above device_type/platform will set the port as 23.


Regards,

Kirk

**Comment by bgile** [comment]
I was doing some more testing and I think there is an issue with the python version running on my test box.  You have as your poetry dependency python = "^3.6" and my test box is using 3.6.8 for its pipenv but when running the same code on a different test box with 3.8 for the pipenv I have no issues.  Not sure if this holds any weight.  I am going to look into getting my test box moved to 3.8 so I can verify that indeed 3.6.8 is the root issue. 

Otherwise, connection_options seems to be working as intended as you stated originally.

**Comment by bgile** [comment]
Well, I retract my statement earlier.  I thought I had this working in a pipenv that was using 3.8 but apparently my testing was incorrect.  I can not get this code to work in any pipenv that I set up.  No matter what I am getting `netmiko.ssh_exception.NetmikoAuthenticationException: Login failed:` for a telnet login.  I'm not certain of the degree you really want to pursue.  When dealing with virtualenv there is a lot of other vars that can affect what is happening.

Before I moved to nornir version 3 and these updated plugins this all worked fine.  If you want to continue looking at this I'm willing to continue getting any info, otherwise, I can understand if it is not worth the effort.  I am going to continue to see if I can work around this in any way.
